### PR TITLE
Allocate default killCmd structs in newE2eService function

### LINKED ISSUE #27582: node e2e test teardown crashes with nil deref if node services could not be started
In `/test/e2e_node/e2e_node_suite_test.go`, the node services are started in the `BeforeSuite` logic. If they fail to start, ginkgo's `Fail` function is called, which ultimately initiates test tear down and thus the `AfterSuite` logic in the same file.

The `AfterSuite` logic calls `e2es.stop()` (defined in `test/e2e_node/e2e_service.go`), which will try to e.g. `es.stopService(es.kubeletCmd)`, where `es` is the `e2eService` and `kubeletCmd` is the `*killCmd` for the kubelet service. 

Unfortunately, because the service failed to start, this pointer is nil, and we get a nil pointer dereference in the `Kill` function (in `test/e2e_node/e2e_service.go`). 

Off the top of my head, I see two potential solutions to this problem.

1. Pre-allocate default `killCmd` structs on the `e2eService` struct created in the `newE2eService` function in `test/e2e_node/e2e_service.go`. We could just make `killCmd.cmd = nil` on these, because the assumption appears to be that if `killCmd.cmd == nil`, the service is not running. (It appears that prior to the addition of `killCmd`, a nil `cmd` was interpreted as 'service not running,' and that this `cmd` field got rolled into `killCmd`, see 66d74f969dbbde894949b87366d215e020167faf).
2. Or, simply add an additional check to the `Kill` function, where we assume that if the `killCmd` is `nil`, the respective service is not running.

Thoughts? I want to make sure I'm not missing something here.

@timstclair because you added teardown code in 66d74f969dbbde894949b87366d215e020167faf
@euank because you added `killCmd` in 34996f1b84104a83033fb62cec72ec86f9eef8c1

Note: Tim is on vacation so I'm not really expecting a response from him.

### PR DESCRIPTION
**Pull request by mtaufen**
Fixes #27582 

[Analytics]()

### CODE REVIEW

**Comment by timstclair** [comment]
instead, can you just add a nil check in `stopService`?

**Comment by mtaufen** [comment]
@timstclair Seems like you and @euank have slightly different preferences on this (see #27582). Do y'all have concrete reasons for choosing one over the other? I gave a somewhat fragile reason for the existing solution in the #27582 discussion. I'm interested in hearing your opinions. Let's not spend a whole lot of time arguing this though :P.

**Comment by euank** [comment]
I don't feel strongly, I just think that avoiding `nil` as much as possible reduces the chance of more accidental NPEs in the future.

LGTM as is, but pending @timstclair's opinion if he'd like it to be the other way.

**Comment by timstclair** [comment]
I agree that nil's are bad, but that's the closest we have for an optional value in go. I prefer a nil in this case because it forces us to explicitly handle a programming error (e.g. program crashes if we don't handle it, rather than mysterious behavior), and it makes the meaning of the value clear. Also I think the nil handling solution is simpler.

Either way is fine, sorry I missed the discussion on the issue. Michael you should choose your favorite :)

**Comment by mtaufen** [comment]
Given that we don't have a compile-time exhaustiveness check for handling optional values, it seems like runtime NPEs are the best we have. It would be interesting if we could spec values as optional and have the testing infrastructure generate tests that just stick `nil` in there to see what happens...

I thought about this for another half hour, and I don't want to split hairs any further. Here's my reasoning:

(A): If you can only discover a failure to exhaustively handle an optional value at runtime, it makes sense to use as few optional values as possible, thereby reducing chance for NPEs. 

(B):  It seems like there are only two 'optional' concepts here: (a) the service was started and (b) an `override` was provided. Given (A), it would be cleanest to only do two checks for `nil`, if possible:
- A check for `killCmd == nil`, because at present this is what happens when there is an error starting a service (as presently written, `cmd` will not be `nil` on the returned `killCmd`, even though the service failed to start). 
- A check for `killCmd.override == nil`

(C): Unfortunately, as long as any of these things are pointers, we need to do a `nil` check to avoid NPEs. 

(D): `killCmd` will probably only ever be used in teardown code, I don't anticipate a _ton_ of reuse of this struct, and therefore we probably don't need to worry about creating a significant "remember to check for `nil`" burden.

Given (C) and (D), I think it makes sense to check all three values (`killCmd`, `override`, and `cmd`), in spite of (A) and (B), so I'm going to go with @timstclair on this.

**Comment by mtaufen** [comment]
Ugh, there's another consideration though, now that `killCmd` == `nil` is the indication that a service is not running, we've lost the ability to say _which_ service isn't running in a log message, because we used to at least have a `name` next to `cmd`.

Maybe we should do something like this along with the additional nil check?:
```
cmd, err = es.startKubeletServer()
es.kubeletCmd = cmd // So we still get a name.
if err != nil {
	es.kubeletCmd.cmd = nil
	return err
}
```

**Comment by timstclair** [comment]
Assuming that an error was logged when the service failed to start, do you really need to it's name again when you say it's not running in teardown?

**Comment by mtaufen** [comment]
That's a good point, I suppose we don't.

**Comment by euank** [comment]
still LGTM

**Comment by mtaufen** [comment]
Thanks Euan :).

**Comment by timstclair** [comment]
LGTM

File: test/e2e_node/e2e_service.go [hunk 1]
```diff
@@ -314,9 +314,15 @@ type killCmd struct {
 }
 
 func (k *killCmd) Kill() error {
+	if k == nil {
+		glog.V(2).Infof("The killCmd is nil, nothing will be killed")
```
Comment (lavalamp) [inline test/e2e_node/e2e_service.go]: I guess there was discussion on this but IMO calling methods on nil pointers is at the very least a bad code smell, and it should be documented very clearly as explicitly allowed.

Comment (lavalamp) [inline test/e2e_node/e2e_service.go]: Like, you could also set the override command to `echo "no need to kill"` in that case...

**Pull request merged**